firewood/src/merkle/stream.rs

```diff
@@ -1,47 +1,228 @@
 // Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
 // See the file LICENSE.md for licensing terms.
 
-use super::{node::Node, BranchNode, Merkle, NodeObjRef, NodeType};
+use super::{node::Node, BranchNode, Merkle, NodeType};
 use crate::{
+    nibbles::Nibbles,
     shale::{DiskAddress, ShaleStore},
     v2::api,
 };
 use futures::{stream::FusedStream, Stream};
-use helper_types::{Either, MustUse};
 use std::task::Poll;
-
 type Key = Box<[u8]>;
 type Value = Vec<u8>;
 
-enum IteratorState<'a> {
-    /// Start iterating at the specified key
-    StartAtKey(Key),
-    /// Continue iterating after the last node in the `visited_node_path`
-    Iterating {
-        visited_node_path: Vec<(NodeObjRef<'a>, u8)>,
+/// Represents an ongoing iteration over a branch node's children.
+struct BranchIterator {
+    /// The nibbles of the key at this branch node.
+    key_nibbles: Vec<u8>,
+    /// Returns the non-empty children of this node and their positions
+    /// in the node's children array .
+    children_iter: Box<dyn Iterator<Item = (DiskAddress, u8)> + Send>,
+}
+
+enum IteratorState {
+    /// The iterator state is lazily initialized when poll_next is called
+    /// for the first time. The iteration start key is stored here.
+    Uninitialized(Key),
+    Initialized {
+        /// Each element is an iterator over a branch node we've visited
+        /// along our traversal of the key-value pairs in the trie.
+        /// We pop an iterator off the stack and call next on it to
+        /// get the next child node to visit. When an iterator is empty,
+        /// we pop it off the stack and go back up to its parent.
+        branch_iter_stack: Vec<BranchIterator>,
     },
 }
 
-impl IteratorState<'_> {
+impl IteratorState {
     fn new() -> Self {
-        Self::StartAtKey(vec![].into_boxed_slice())
+        Self::Uninitialized(vec![].into_boxed_slice())
     }
 
     fn with_key(key: Key) -> Self {
-        Self::StartAtKey(key)
+        Self::Uninitialized(key)
     }
 }
 
 /// A MerkleKeyValueStream iterates over keys/values for a merkle trie.
 pub struct MerkleKeyValueStream<'a, S, T> {
-    key_state: IteratorState<'a>,
+    key_state: IteratorState,
     merkle_root: DiskAddress,
     merkle: &'a Merkle<S, T>,
 }
 
+/// Returns the state of an iterator over merkle, which has the given
+/// root_node, after it's been initialized. Iteration starts at the given key.
+fn get_iterator_intial_state<S: ShaleStore<Node> + Send + Sync, T>(
+    merkle: &Merkle<S, T>,
+    root_node: DiskAddress,
+    key: &[u8],
+) -> Result<IteratorState, api::Error> {
+    let root_node = merkle
+        .get_node(root_node)
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
```

Comment:
The error handling is poor here; if there is an IO error it gets mapped to an internal error. The right fix for this is to write `From` implementations for `MerkleError` and `ShaleError` for `v2::api::Error`.

This should be done in a follow up though.